src/core/cgroup.c

```diff
@@ -3510,6 +3510,34 @@ void manager_invalidate_startup_units(Manager *m) {
                 unit_invalidate_cgroup(u, CGROUP_MASK_CPU|CGROUP_MASK_IO|CGROUP_MASK_BLKIO);
 }
 
+int compare_job_cpu_weight(const void *a, const void *b) {
+        const Job *x = a, *y = b;
+        ManagerState state = manager_state(x->unit->manager);
+        CGroupContext *context;
+        uint64_t weight_x, weight_y;
+        int ret;
+
+        context = unit_get_cgroup_context(x->unit);
+        weight_x = context ? cgroup_context_cpu_weight(context, state) : CGROUP_WEIGHT_DEFAULT;
+        context = unit_get_cgroup_context(y->unit);
+        weight_y = context ? cgroup_context_cpu_weight(context, state) : CGROUP_WEIGHT_DEFAULT;
+
+        /* Sort jobs for all units by CPU weight:
+         * - units that fork new processes after any other units
+         * - units with higher CPU weight first
+         * - sort by id if the other criteria are the same */
+        if (x->unit->type > UNIT_SWAP && y->unit->type <= UNIT_SWAP)
+                return -1;
+
+        if (x->unit->type <= UNIT_SWAP && y->unit->type > UNIT_SWAP)
+                return 1;
```

Comment:
not sure I get this, I thought we agreed on just making this like this:

```c
ret = CMP(x->unit->type, y->unit->type);
```

With that in place the order in the unit type enum would be sufficient to order and we need nothing else.

Reply:
hmm, another idea might be to use the nice level for ordering too, before resorting to the cpu weight, but doesn't matter much

Reply:
With just a `CMP` the unit type has the highest priority. This means that any service is always started before (or after) a mount unit regardless of the cpu weight.
I don't think that's a good idea.

I'll take a look at the nice level.

Reply:
I've implemented the nice level handling.
I'd like to keep the type comparison as it is. If you prefer it otherwise then I think the following priority order would make sense:
target -> * -> socket -> service -> mount -> swap
Does that make sense? I don't really care about most of the order, but I think services should have a higher priority than mount or swap units.

Reply:
nah, i would say: order by nice level first. if both units have the same nice level, use cpu weight. if both have the same cpu weight, then user unit type with a simple CMP, then strcmp() of the name. This means that the nice level always is most important, and the type only as a fallback

Reply:
So, I've experimented some more. If I implement this as suggested then the following happens:
Lets say we have:
- a.target
- b.target after a.target
- a.service after a.target with cpu weight 200
- b.service after b.target with cpu weight 300

Because b.target gets the default cpu weight, it is sorted after a.service. As a result, b.service is started after a.service.
The same happens with b.socket instead of b.target and that is a very common pattern. At least here I can set the cpu weight for the socket as well but imho that's not very intuitive.

If this is acceptable behavior, then I can update the PR like this. But I think it would make sense to drop the unit type sorting entirely. It does not add any value. I only introduced it to avoid cases like the example above.

Reply:
well, the After= and Before= ordering always takes precedence, obviously, but I think there's still benefit in scheduling services before targets when both are runnable at the same time, simply to increase parallelism. Hence I would still prefer services over targers, and only as last resort compare unit names.

Reply:
> At least here I can set the cpu weight for the socket as well but imho that's
> not very intuitive.
It is good that user can tweak it (and that the default is somewhat
predictable).

An idea for the units without execution context -- what about using
init.scope's CPUWeight= instead of the hardwired default? (Note that without effective weight, it's apples to oranges comparison, so the behavior should be documented for fine-tuning users.)

Reply:
@poettering ok, I think I've implemented it as requested now.